File: 01-Languages/CSharp/21-Solutions/solution-07.cs
// solution-07.cs - JSON Roundtrip with LINQ Filtering
// See: ../20-Exercises/README.md#exercise-07--json-roundtrip-with-linq-filtering-advanced
//
// Run with:
//     dotnet run solution-07.cs

using System.Text.Json;

// .NET 10 file-based apps disable reflection-based JSON serialization by
// default (it's tuned for trimming/AOT scenarios) -- without this switch,
// JsonSerializer.Serialize below throws InvalidOperationException at
// runtime even though the code compiles cleanly. See Lesson 10's example.cs
// for the first place this course hits the same quirk.
AppContext.SetSwitch("System.Text.Json.JsonSerializer.IsReflectionEnabledByDefault", true);

var books = new List<Book> {
    new("Clean Code", "Robert C. Martin", 2008, 4.2),
    new("The Pragmatic Programmer", "Hunt & Thomas", 1999, 4.4),
    new("C# in Depth", "Jon Skeet", 2019, 4.7),
    new("Deep Work", "Cal Newport", 2016, 4.3),
    new("Atomic Habits", "James Clear", 2018, 4.8),
    new("Refactoring", "Martin Fowler", 1999, 4.5),
};

// Path.GetTempFileName() guarantees a unique, writable path so this
// example never collides with a real file or needs manual name generation.
var tempPath = Path.GetTempFileName();

try {
    Console.WriteLine($"--- serializing {books.Count} books to {tempPath} ---");
    var json = JsonSerializer.Serialize(books, new JsonSerializerOptions { WriteIndented = true });
    File.WriteAllText(tempPath, json);

    Console.WriteLine("\n--- reading back and deserializing ---");
    var readBack = File.ReadAllText(tempPath);
    var deserialized = JsonSerializer.Deserialize<List<Book>>(readBack)
        ?? throw new InvalidOperationException("Deserialization returned null.");
    Console.WriteLine($"Deserialized {deserialized.Count} books.");

    Console.WriteLine("\n--- books after 2015 with rating >= 4.0, sorted by rating desc ---");
    var filtered = deserialized
        .Where(b => b.Year > 2015 && b.Rating >= 4.0)
        .OrderByDescending(b => b.Rating);
    foreach (var b in filtered) {
        Console.WriteLine($"  {b.Title} ({b.Year}) by {b.Author} -- {b.Rating:F1}");
    }
} finally {
    // Cleanup runs even if an assertion above throws, so a failed run
    // never leaves a stray temp file behind.
    File.Delete(tempPath);
    Console.WriteLine($"\nCleanup check -- file still exists: {File.Exists(tempPath)}");
}

record Book(string Title, string Author, int Year, double Rating);

File: 01-Languages/Swift/21-Solutions/solution-07.swift
// solution-07.swift -- Exercise 07: Codable JSON Roundtrip with Validation

import Foundation

struct Book: Codable, Equatable {
    let title: String
    let author: String
    let year: Int
    let rating: Double
}

enum BookValidationError: Error {
    case ratingOutOfRange(Double)
}

func validate(_ book: Book) throws -> Book {
    guard (0.0...5.0).contains(book.rating) else {
        throw BookValidationError.ratingOutOfRange(book.rating)
    }
    return book
}

let candidates = [
    Book(title: "Clean Code", author: "Robert C. Martin", year: 2008, rating: 4.5),
    Book(title: "The Pragmatic Programmer", author: "Hunt & Thomas", year: 1999, rating: 4.8),
    Book(title: "Impossible Ratings", author: "Nobody", year: 2020, rating: 7.5), // invalid
    Book(title: "Refactoring", author: "Martin Fowler", year: 1999, rating: 4.6),
]

var validBooks: [Book] = []

print("--- Validation ---")
for book in candidates {
    do {
        let validated = try validate(book)
        validBooks.append(validated)
        print("OK: \(book.title)")
    } catch BookValidationError.ratingOutOfRange(let rating) {
        // Invalid data is caught HERE and never reaches the encoding step below --
        // `validBooks` only ever accumulates books that passed `validate(_:)`.
        print("REJECTED: \(book.title) has an out-of-range rating (\(rating)); not encoded.")
    } catch {
        print("Unexpected error for \(book.title): \(error)")
    }
}

print("\n--- Encode valid books to JSON ---")
let encoder = JSONEncoder()
encoder.outputFormatting = .prettyPrinted
let jsonData = try! encoder.encode(validBooks)
let jsonString = String(data: jsonData, encoding: .utf8)!
print(jsonString)

print("\n--- Decode back and confirm roundtrip equality ---")
let decoder = JSONDecoder()
let roundtripped = try! decoder.decode([Book].self, from: jsonData)
print("roundtripped == validBooks: \(roundtripped == validBooks)")
